enclave_build/src/image.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{EnclaveBuildError, Result};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::io::Read;
+
+use sha2::Digest;
+
+use oci_distribution::{client::ImageData, Reference};
+use oci_spec::image::ImageConfiguration;
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Debug)]
+/// Struct representing the image metadata, like the image ID (hash) and config.
+pub struct ImageDetails {
+    /// The reference of an image, e.g. "docker.io/library/hello-world:latest".
+    // Use a String, since the oci_distribution::Reference struct does not implement
+    // Serialize/Deserialize
+    pub uri: String,
+    /// The image ID, calculated as the SHA256 digest hash of the image config.
+    #[serde(rename = "Id")]
+    pub hash: String,
+    /// The image config.
+    pub config: ImageConfiguration,
+}
+
+impl ImageDetails {
+    pub fn new<S: AsRef<str>>(image_uri: S, image_hash: S, config: ImageConfiguration) -> Self {
+        Self {
+            uri: image_uri.as_ref().to_string(),
+            hash: image_hash.as_ref().to_string(),
+            config,
+        }
+    }
+
+    /// Try to build an ImageDetails struct from an oci_distribution ImageData struct.
+    //
+    // The oci_distribution ImageData struct does not contain the image name or reference, so this
+    // must be additionally passed to the function as well.
+    pub fn from<S: AsRef<str>>(image_name: S, image_data: &ImageData) -> Result<Self> {
+        // Get the config JSON String from the pulled image
+        let config_json = String::from_utf8(image_data.config.data.clone())
+            .map_err(|err| EnclaveBuildError::ImageDetailError(format!("{:?}", err)))?;
+
+        // Calculate the image hash as the digest of the image config, as specified in the OCI image spec
+        // https://github.com/opencontainers/image-spec/blob/main/config.md
+        let image_hash = format!("sha256:{:x}", sha2::Sha256::digest(config_json.as_bytes()));
+
+        let image_ref = build_image_reference(&image_name)?;
+
+        Ok(Self {
+            uri: image_ref.whole(),
+            hash: image_hash,
+            config: deserialize_from_reader(config_json.as_bytes())?,
+        })
+    }
+}
+
+/// Calculates the image ID (or image hash) as the SHA256 digest of the image config.
+///
+/// This method is described in the OCI image spec.
+///
+/// https://github.com/opencontainers/image-spec/blob/main/config.md
+pub fn image_hash<R: Read>(mut image_config: R) -> Result<String> {
```

Comment:
This method seems to be a bit useless. We are using it only with slice parameters, for which we can just do the `format!` below, so no need to have a new Vec and move data around.